File: backend/app/schemas/appointment.py
"""Appointment document schema for frontend alignment."""
from datetime import datetime, timezone
from typing import Optional

from pydantic import BaseModel, Field

from .objectid import PyObjectId


def _utc_now() -> datetime:
    return datetime.now(timezone.utc)


class AppointmentDoc(BaseModel):
    """Mongo document shape for appointments collection."""

    id: Optional[PyObjectId] = Field(None, alias="_id")
    patient_id: PyObjectId
    doctor_id: PyObjectId
    scheduled_at: datetime
    status: str = "scheduled"  # scheduled | completed | cancelled | no_show
    reason: str = ""
    notes: Optional[str] = None
    created_at: datetime = Field(default_factory=_utc_now)
    updated_at: datetime = Field(default_factory=_utc_now)

    model_config = {"populate_by_name": True, "arbitrary_types_allowed": True}

File: express-backend/src/models/Appointment.ts
import mongoose, { Schema, type Document, type Model } from 'mongoose';

export interface AppointmentDocument extends Document {
  patient_id: mongoose.Types.ObjectId;
  doctor_id: mongoose.Types.ObjectId;
  proposed_slots: Date[];
  scheduled_at?: Date | null;
  requested_by_role: 'patient' | 'doctor';
  status: string;
  reason: string;
  notes?: string;
  response_message?: string;
  created_at: Date;
  updated_at: Date;
}

const AppointmentSchema = new Schema<AppointmentDocument>(
  {
    patient_id: { type: Schema.Types.ObjectId, ref: 'Patient', required: true },
    doctor_id: { type: Schema.Types.ObjectId, ref: 'Doctor', required: true },
    proposed_slots: [{ type: Date, required: true }],
    scheduled_at: { type: Date, default: null },
    requested_by_role: {
      type: String,
      enum: ['patient', 'doctor'],
      required: true,
    },
    status: {
      type: String,
      enum: ['pending', 'confirmed', 'rejected', 'cancelled', 'completed'],
      default: 'pending',
    },
    reason: { type: String, default: '' },
    notes: { type: String },
    response_message: { type: String },
    created_at: { type: Date, default: () => new Date() },
    updated_at: { type: Date, default: () => new Date() },
  },
  {
    collection: 'appointments',
  },
);

AppointmentSchema.pre('save', function (next) {
  this.updated_at = new Date();
  next();
});

export const Appointment: Model<AppointmentDocument> =
  mongoose.models.Appointment ||
  mongoose.model<AppointmentDocument>('Appointment', AppointmentSchema);

